README.md
# Kusk CLI

[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=shields)](http://makeapullrequest.com)

Kusk is a CLI tool designed to help you manage common tasks required when running Kusk Gateway.

Currently we support the following commands:

- `install` - installs Kusk Gateway and all its components with a single command. (Requires a helm installation)
- `api generate` - for creating Kusk Gateway API resources from your OpenAPI specification document.

---

# Table of contents

- [Usage](#usage)
  - [install](#install)
  - [api generate](#api-generate)
  - [dashboard](#dashboard)
- [Installation](#installation)
- [Updating](#updating)
- [Uninstallation](#uninstallation)
- [Contributing](#contributing)
- [License](#license)

# Usage

## Install

### Flags

|         Flag         |                                                     Description                                                     | Required?  |
|:--------------------:|:-------------------------------------------------------------------------------------------------------------------:|:----------:|
|       `--name`       | the prefix of the name to give to the helm releases for each of the kusk gateway components (default: kusk-gateway) |     ❌     |
| `--namespace` / `-n` |   the namespace to install kusk gateway into. Will create the namespace if it doesn't exist (default: kusk-system)  |     ❌     |
|   `--no-dashboard`   |                               when set, will not install the kusk gateway dashboard.                                |     ❌     |
|      `--no-api`      |                       when set, will not install the kusk gateway api. implies --no-dashboard.                      |     ❌     |
|  `--no-envoy-fleet`  |                                     when set, will not install any envoy fleets                                     |     ❌     |

### Examples

```sh
$ kusk install

Will install kusk-gateway, a public (for your APIS) and private (for the kusk dashboard and api)
envoy-fleet, api, and dashboard in the kusk-system namespace using helm.

$ kusk install --name=my-release --namespace=my-namespace

Will create a helm release named with --name in the namespace specified by --namespace.

$ kusk install --no-dashboard --no-api --no-envoy-fleet

Will install kusk-gateway, but not the dashboard, api, or envoy-fleet.
```

## Api generate

Generate accepts your OpenAPI spec file as input either as a local file or a URL pointing to your file
and generates a Kusk Gateway compatible API resource that you can apply directly into your cluster.

Configuration of the API resource is done via the x-kusk extension.

If the OpenAPI spec doesn't have a top-level x-kusk annotation set, it will add them for you and set
the upstream service, namespace and port to the flag values passed in respectively and set the rest of the settings to defaults.
This is enough to get you started

If the x-kusk extension is already present, it will override the the upstream service, namespace and port to the flag values passed in respectively
and leave the rest of the settings as they are.

You must specify the name of the envoyfleet you wish to use to expose your API. This is because Kusk Gateway could be managing more than one.
In the future, we will add the notion of a default envoyfleet which kusk gateway will use when none is specified. i.e. kusk-gateway-envoy-fleet.

If you do not specify the envoyfleet namespace, it will default to kusk-system.

Sample usage

No name specified

```sh
kusk api generate \
  -i spec.yaml \
  --envoyfleet.name kusk-gateway-envoy-fleet \
  --envoyfleet.namespace kusk-system
```

In the above example, kusk will use the openapi spec info.title to generate a manifest name and leave the existing
x-kusk extension settings

No api namespace specified

```sh
kusk api generate \
  -i spec.yaml \
  --name httpbin-api \
  --upstream.service httpbin \
  --upstream.port 8080 \
  --envoyfleet.name kusk-gateway-envoy-fleet
```

In the above example, as --namespace isn't defined, it will assume the default namespace.

Namespace specified

```sh
kusk api generate \
  -i spec.yaml \
  --name httpbin-api \
  --upstream.service httpbin \
  --upstream.namespace my-namespace \
  --upstream.port 8080 \
  --envoyfleet.name kusk-gateway-envoy-fleet
```

OpenAPI spec at URL

```sh
kusk api generate \
    -i https://raw.githubusercontent.com/$ORG_OR_USER/$REPO/myspec.yaml \
    --name httpbin-api \
    --upstream.service httpbin \
    --upstream.namespace my-namespace \
    --upstream.port 8080 \
    --envoyfleet.name kusk-gateway-envoy-fleet
```

This will fetch the OpenAPI document from the provided URL and generate a Kusk Gateway API resource

### Flags

|          Flag          |                                             Description                                             | Required? |
|:----------------------:|:---------------------------------------------------------------------------------------------------:|:---------:|
|        `--name`        | the name to give the API resource e.g. --name my-api. Otherwise taken from OpenAPI info title field |     ❌     |
|  `--namespace` / `-n`  | the namespace of the API resource e.g. --namespace my-namespace, -n my-namespace (default: default) |     ❌     |
|      `--in` / `-i`     |       file path or URL to OpenAPI spec file to generate mappings from. e.g. --in apispec.yaml       |     ✅     |
|  `--upstream.service`  |                                 name of upstream Kubernetes service                                 |     ❌     |
| `--upstream.namespace` |                           namespace of upstream service (default: default)                          |     ❌     |
|    `--upstream.port`   |                        port that upstream service is exposed on (default: 80)                       |     ❌     |
|   `--envoyfleet.name`  |                                name of envoyfleet to use for this API                               |     ✅     |
| `envoyfleet.namespace` |                  namespace of envoyfleet to use for this API. Default: kusk-system                  |     ❌     |

### Example

Take a look at the [`./examples/httpbin-spec.yaml`](./examples/httpbin-spec.yaml)

```sh
kusk api generate -i ./examples/httpbin-spec.yaml --name httpbin-api --upstream.service httpbin --upstream.port 8080 --envoyfleet.name kusk-gateway-envoy-fleet
```

The output should contain the following x-kusk extension at the top level

```sh
...
x-kusk:
  cors: {}
  path:
    rewrite:
      pattern: ""
      substitution: ""
  upstream:
    service:
	name: httpbin
	namespace: default
	port: 8080
```

## dashboard
Access the kusk dashboard. kusk dashboard will start a port-forward session on port 8080 to the envoyfleet
serving the dashboard and will open the dashboard in the browser. By default this is kusk-gateway-private-envoy-fleet.kusk-system.

The flags --envoyfleet.namespace and --envoyfleet.name can be used to change the envoyfleet.

### Flags
|           Flag           |                                          Description                                         | Required? |
|:------------------------:|:--------------------------------------------------------------------------------------------:|:---------:|
|      `--kubeconfig`      |                                 absolute path to kube config                                 |     ❌     |
|    `--envoyfleet.name`   | kusk gateway dashboard envoy fleet service name. (default: kusk-gateway-private-envoy-fleet) |     ❌     |
| `--envoyfleet.namespace` |         kusk gateway dashboard envoy fleet service namespace. (default: kusk-system)         |     ❌     |
|     `--external-port`    |                     external port to access dashboard at. (default: 8080)                    |     ❌     |

### Examples
```
$ kusk dashboard
```

Opens the kusk gateway dashboard in the browser by exposing the default private envoy fleet on port 8080

```
$ kusk dashboard --envoyfleet.namespace=other-namespace --envoyfleet.name=other-envoy-fleet
```

Specify other envoyfleet and namespace that is serving the dashboard

```
$ kusk dashboard --external-port=9090
```

Expose dashboard on port 9090

```
$ kusk dashboard --kubeconfig=/path/to/kube/config
```
Specify path to kube config. $HOME/.kube/config is used by default.

# Installation

## Homebrew

```sh
brew install kubeshop/kusk/kusk
```

## Go install the latest release on Github

```sh
go install github.com/kubeshop/kusk@latest
```

To install a particular version: replace `latest` with the version number

You can get a list of the available kusk versions from our [releases page](https://github.com/kubeshop/kusk/releases)

## Easy install script

This will install `kusk` into `/usr/local/bin/kusk`

```sh
$ curl -sSLf https://raw.githubusercontent.com/kubeshop/kusk/main/scripts/install.sh | sh
...
kusk installed in /usr/local/bin/kusk
```

## From source

```sh
git clone git@github.com:kubeshop/kusk.git && \
  cd kusk && \
  go install
```

## Alternative installation method (manual)

If you don't like automatic scripts you can always use the manual install:

1. Download binary with version of your choice (recent one is recommended).
2. Unpack it (`tar -zxvf kusk_0.1.0_Linux_arm64.tar.gz`).
3. Move it to a location in the `PATH`. For example `mv kusk_0.1.0_Linux_arm64/kusk /usr/local/bin/kusk`.

For Windows, download the binary from [here](https://github.com/kubeshop/kusk/releases), unpack the binary and add it to `%PATH%`.

# Updating

## Homebrew

```sh
brew upgrade kubeshop/kusk/kusk
```

## Latest release on Github

```
go install github.com/kubeshop/kusk@latest
```

## From source

Inside of the kusk repository directory

```sh
git pull && go install
```

# Contributing

Your contributions are always welcome! Please have a look at [How to contribute](https://github.com/kubeshop/.github/blob/main/CONTRIBUTING.md) first.

### Development

```sh
make
```

# License

[The MIT License](https://mit-license.org/).

Please have a look at the [`./LICENSE`](./LICENSE) for more details.


vscode/CHANGELOG.md
# Change Log

All notable changes to the "kusk" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

cmd/install.go
/*
The MIT License (MIT)

Copyright © 2022 Kubeshop

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

*/
package cmd

import (
	"fmt"
	"os/exec"
	"strings"

	"github.com/kubeshop/testkube/pkg/process"
	"github.com/kubeshop/testkube/pkg/ui"
	"github.com/spf13/cobra"
)

var (
	noApi                         bool
	noDashboard                   bool
	noEnvoyFleet                  bool
	releaseName, releaseNamespace string
)

func init() {
	rootCmd.AddCommand(installCmd)
	installCmd.Flags().StringVar(&releaseName, "name", "kusk-gateway", "installation name")
	installCmd.Flags().StringVar(&releaseNamespace, "namespace", "kusk-system", "namespace to install in")

	installCmd.Flags().BoolVar(&noDashboard, "no-dashboard", false, "don't the install dashboard")
	installCmd.Flags().BoolVar(&noApi, "no-api", false, "don't install the api. Setting this flag implies --no-dashboard")
	installCmd.Flags().BoolVar(&noEnvoyFleet, "no-envoy-fleet", false, "don't install any envoy fleets")
}

var installCmd = &cobra.Command{
	Use:   "install",
	Short: "Install kusk-gateway, envoy-fleet, api, and dashboard in a single command",
	Long: `
	Install kusk-gateway, envoy-fleet, api, and dashboard in a single command.

	$ kusk install

	Will install kusk-gateway, a public (for your APIS) and private (for the kusk dashboard and api) 
	envoy-fleet, api, and dashboard in the kusk-system namespace using helm.

	$ kusk install --name=my-release --namespace=my-namespace

	Will create a helm release named with --name in the namespace specified by --namespace.

	$ kusk install --no-dashboard --no-api --no-envoy-fleet

	Will install kusk-gateway, but not the dashboard, api, or envoy-fleet.
	`,
	Run: func(cmd *cobra.Command, args []string) {
		helmPath, err := exec.LookPath("helm")
		ui.ExitOnError("looking for helm", err)

		ui.Info("adding the kubeshop helm repository")
		err = addKubeshopHelmRepo(helmPath)
		ui.ExitOnError("adding kubeshop repo", err)
		ui.Info(ui.Green("done"))

		ui.Info("fetching the latest charts")
		err = updateHelmRepos(helmPath)
		ui.ExitOnError("updating helm repositories", err)
		ui.Info(ui.Green("done"))

		ui.Info("installing Kusk Gateway")
		err = installKuskGateway(helmPath, releaseName, releaseNamespace)
		ui.ExitOnError("Installing kusk gateway", err)
		ui.Info(ui.Green("done"))

		if !noEnvoyFleet {
			ui.Info("installing Envoy Fleet")
			err = installPublicEnvoyFleet(helmPath, releaseName, releaseNamespace)
			ui.ExitOnError("Installing envoy fleet", err)
			ui.Info(ui.Green("done"))
		} else {
			ui.Info(ui.LightYellow("--no-envoy-fleet set - skipping envoy fleet installation"))
		}

		if noApi {
			ui.Info(ui.LightYellow("--no-api set - skipping api installation"))
			return
		}

		ui.Info("installing Kusk API")
		envoyFleetName := fmt.Sprintf("%s-envoy-fleet", releaseName)
		if !noEnvoyFleet {
			envoyFleetName = fmt.Sprintf("%s-private-envoy-fleet", releaseName)
			err = installPrivateEnvoyFleet(helmPath, fmt.Sprintf("%s-private", releaseName), releaseNamespace)
			ui.ExitOnError("Installing envoy fleet", err)
		}

		err = installApi(helmPath, releaseName, releaseNamespace, envoyFleetName)
		ui.ExitOnError("Installing api", err)
		ui.Info(ui.Green("done"))

		if noDashboard {
			ui.Info(ui.LightYellow("--no-dashboard set - skipping dashboard installation"))
			printPortForwardInstructions("API", releaseNamespace, envoyFleetName)
			return
		}

		ui.Info("installing Kusk Dashboard")
		err = installDashboard(helmPath, releaseName, releaseNamespace, envoyFleetName)
		ui.ExitOnError("Installing dashboard", err)

		ui.Info(ui.Green("done"))

		printPortForwardInstructions("dashboard", releaseNamespace, envoyFleetName)
	},
}

func printPortForwardInstructions(service, releaseNamespace, envoyFleetName string) {
	ui.Info(ui.Green("To access the " + service + ", port forward to the envoy-fleet service that exposes it"))
	ui.Info(ui.LightBlue(fmt.Sprintf("\t$ kubectl port-forward -n %s svc/%s 8080:80", releaseNamespace, envoyFleetName)))

	endpoint := "http://localhost:8080/"
	if service == "api" {
		endpoint += "api"
	}
	ui.Info(ui.LightBlue("\tand go " + endpoint))
}

func addKubeshopHelmRepo(helmPath string) error {
	_, err := process.Execute(helmPath, "repo", "add", "kubeshop", "https://kubeshop.github.io/helm-charts/")
	if err != nil && !strings.Contains(err.Error(), "Error: repository name (kubeshop) already exists, please specify a different name") {
		return err
	}

	return nil
}

func updateHelmRepos(helmPath string) error {
	_, err := process.Execute(helmPath, "repo", "update")
	return err
}

func installKuskGateway(helmPath, releaseName, releaseNamespace string) error {
	command := []string{
		"upgrade",
		"--install",
		"--wait",
		"--create-namespace",
		"--namespace",
		releaseNamespace,
		"--set", fmt.Sprintf("fullnameOverride=%s", releaseName),
		releaseName,
		"kubeshop/kusk-gateway",
	}

	out, err := process.Execute(helmPath, command...)
	if err != nil {
		return err
	}

	ui.Debug("helm install kusk gateway output", string(out))

	return nil
}

func installPublicEnvoyFleet(helmPath, releaseName, releaseNamespace string) error {
	return installEnvoyFleet(helmPath, releaseName, releaseNamespace, "LoadBalancer")
}

func installPrivateEnvoyFleet(helmPath, releaseName, releaseNamespace string) error {
	return installEnvoyFleet(helmPath, releaseName, releaseNamespace, "ClusterIP")
}

func installEnvoyFleet(helmPath, releaseName, releaseNamespace, serviceType string) error {
	envoyFleetName := fmt.Sprintf("%s-envoy-fleet", releaseName)
	command := []string{
		"upgrade",
		"--install",
		"--wait",
		"--create-namespace",
		"--namespace",
		releaseNamespace,
		"--set", fmt.Sprintf("fullnameOverride=%s", envoyFleetName),
		"--set", fmt.Sprintf("service.type=%s", serviceType),
		envoyFleetName,
		"kubeshop/kusk-gateway-envoyfleet",
	}

	out, err := process.Execute(helmPath, command...)
	if err != nil {
		return err
	}

	ui.Debug("helm install envoy fleet output", string(out))

	return nil
}

func installApi(helmPath, releaseName, releaseNamespace, envoyFleetName string) error {
	command := []string{
		"upgrade",
		"--install",
		"--wait",
		"--create-namespace",
		"--namespace",
		releaseNamespace,
		"--set", fmt.Sprintf("fullnameOverride=%s-api", releaseName),
		"--set", fmt.Sprintf("envoyfleet.name=%s", envoyFleetName),
		"--set", fmt.Sprintf("envoyfleet.namespace=%s", releaseNamespace),
		fmt.Sprintf("%s-api", releaseName),
		"kubeshop/kusk-gateway-api",
	}

	out, err := process.Execute(helmPath, command...)
	if err != nil {
		return err
	}

	ui.Debug("helm install api output", string(out))

	return nil
}

func installDashboard(helmPath, releaseName, releaseNamespace, envoyFleetName string) error {
	command := []string{
		"upgrade",
		"--install",
		"--wait",
		"--create-namespace",
		"--namespace",
		releaseNamespace,
		"--set", fmt.Sprintf("fullnameOverride=%s-dashboard", releaseName),
		"--set", fmt.Sprintf("envoyfleet.name=%s", envoyFleetName),
		"--set", fmt.Sprintf("envoyfleet.namespace=%s", releaseNamespace),
		fmt.Sprintf("%s-dashboard", releaseName),
		"kubeshop/kusk-gateway-dashboard",
	}

	out, err := process.Execute(helmPath, command...)
	if err != nil {
		return err
	}

	ui.Debug("helm install dashboard output", string(out))

	return nil
}
